tests/unit/test_salesforce.py

```diff
@@ -0,0 +1,101 @@
+import pandas as pd
+import pytest
+
+from viadot.sources import Salesforce
+
+TABLE_TO_DOWNLOAD = "Account"
+TABLE_TO_UPSERT = "Contact"
+TEST_LAST_NAME = "prefect-viadot-test"
+ID_TO_UPSERT = "0035E00001YGWK3QAP"
+
+
+@pytest.fixture(scope="session")
+def salesforce():
+    s = Salesforce(config_key="sales_force_dev")
+    yield s
+
+
+@pytest.fixture(scope="session")
+def test_df_data(salesforce):
+    data = {
+        "Id": [ID_TO_UPSERT],
+        "LastName": [TEST_LAST_NAME],
+    }
+    df = pd.DataFrame(data=data)
+
+    yield df
+
+    data_restored = {
+        "Id": [ID_TO_UPSERT],
+        "LastName": ["LastName"],
+    }
+    df_restored = pd.DataFrame(data=data_restored)
+    salesforce.upsert(df=df_restored, table=TABLE_TO_UPSERT)
```

Comment:
You should not use a function you're about to test; also the teardown should be cleaning up resources, not creating them

Reply:
[✅ Improved cleaning after tests](https://github.com/dyvenia/viadot/pull/629/commits/b408db7f60c099bc43421eaf070158d13e2a1f8d)

Reply:
The second point remains, I think it may be confusing because you're mixing 2 things together in these fixtures, you're creating a test df but also randomly upserting records. It's actually not easy to understand the tests since you already upsert records in the fixture, I don't think this should be happening? Perhaps you meant to **insert** some records in order to validate  that they are correctly updated in the `Salesforce.upsert()` method? But in that case, I don't think this needs to be in a fixture, you can do this (note also there is a cleanup at the end, pls also add it so the tests do not permanently create/modify any records).
```python
TEST_DF_EXTERNAL = pd.DataFrame({"LastName": [TEST_LAST_NAME], "SAPContactId__c": ["111"]})

def test_upsert(salesforce):
    # Setup.
    sf = salesforce.salesforce
    test_record = TEST_DF_EXTERNAL.to_records()  # not sure if this is the data structure required by insert(), pls check
    sf.Contact.insert(test_record)
     
    # Test.
    salesforce.upsert(
        df=TEST_DF_EXTERNAL, table=TABLE_TO_UPSERT, external_id="SAPContactId__c"
    )

    result = sf.query(
        f"SELECT ID, LastName FROM {TABLE_TO_UPSERT} WHERE ID='{ID_TO_UPSERT}'"
    )
    assert result["records"][0]["LastName"] == TEST_LAST_NAME

    # Cleanup
    sf.delete(test_record)
```